scripts/setup/reindex-textual-data

```diff
@@ -0,0 +1,133 @@
+
+setup_path()
+
+from scripts.lib.zulip_tools import su_to_zulip
+
+su_to_zulip()
+os.environ["DJANGO_SETTINGS_MODULE"] = "zproject.settings"
+from django.conf import settings
+
+parser = argparse.ArgumentParser(
+    description="Reindex all text indexes, for glibc upgrades.  This will take a write lock on every table, unless --concurrently is passed."
+)
+parser.add_argument(
+    "--concurrently", action="store_true", help="reindex concurrently, on Pg ≥ 11; takes longer"
+)
+parser.add_argument("--force", action="store_true", help="run the reindexing")
+options = parser.parse_args()
+
+logging.Formatter.converter = time.gmtime
+logging.basicConfig(format="%(asctime)s %(levelname)s: %(message)s")
+logger = logging.getLogger("reindex-textual-data")
+logger.setLevel(logging.DEBUG)
+
+pg_args = {}
+pg_args["host"] = settings.DATABASES["default"]["HOST"]
+pg_args["port"] = settings.DATABASES["default"].get("PORT")
+pg_args["password"] = settings.DATABASES["default"].get("PASSWORD")
+pg_args["user"] = settings.DATABASES["default"]["USER"]
+pg_args["dbname"] = settings.DATABASES["default"]["NAME"]
+pg_args["sslmode"] = settings.DATABASES["default"]["OPTIONS"].get("sslmode")
+pg_args["connect_timeout"] = "600"
+
+conn = psycopg2.connect(**pg_args)
+conn.autocommit = True
+
+pg_server_version = conn.server_version
+can_concurrently = pg_server_version >= 110000  # Version 11.0.0
+
+if options.concurrently and not can_concurrently:
+    raise RuntimeError("Only PostgreSQL 11 and above can REINDEX CONCURRENTLY.")
+
+cursor = conn.cursor()
+cursor.execute(
+    """
+SELECT
+        irel.relname AS index_name,
+        trel.relname AS table_name,
+        pg_size_pretty(pg_table_size(i.indrelid)) AS size
+FROM
+        pg_index AS i
+        JOIN pg_class AS trel ON trel.oid = i.indrelid
+        JOIN pg_namespace AS tnsp ON trel.relnamespace = tnsp.oid
+        JOIN pg_class AS irel ON irel.oid = i.indexrelid
+        CROSS JOIN unnest(i.indkey) colnum
+        JOIN pg_attribute AS a ON trel.oid = a.attrelid
+                AND a.attnum = colnum
+        WHERE tnsp.nspname = 'zulip'
+        AND a.attcollation != 0
+GROUP BY 1, 2, i.indrelid
+ORDER BY pg_table_size(i.indrelid) ASC
+"""
+)
+reindex_targets = cursor.fetchall()
+
+if not options.force:
+    print("Would reindex the following indexes:")
+    for index, _, _ in reindex_targets:
+        print(f"  {index}")
+    print(
+        """
+
+Re-run with --force to reindex these.  Without --concurrently, these
+index rebuilds will TAKE A WRITE LOCK on their respective tables,
+which means that production traffic will be affected.  On PostgreSQL
+11 and above, you can pass --concurrently, which will do the index
+rebuilds concurrently; however, this comes at a significant
+performance penalty.
+
+"""
+    )
+    sys.exit(0)
+
+had_failures = False
+for index_name, table_name, index_size in reindex_targets:
+    if options.concurrently:
+        command = f"REINDEX INDEX CONCURRENTLY {index_name}"
```

Comment:
For correct SQL interpolation:

```python3
from psycopg2.sql import SQL, Identifier
command = SQL("REINDEX INDEX CONCURRENTLY {}").format(Identifier(index_name))
```

Hopefully doesn’t matter here, but it’s better to be consistently correct, and this will at least fail in a less arbitrary way in case bugs result in unexpected characters in `index_name`.

Reply:
Yeah, I don't think we're realistically worried about SQL-injection-by-malicious-index-name, but it does make sense to provide a good example.  Thanks for the code -- fixed!